core/src/main/java/com/graphhopper/storage/CHGraphImpl.java

```diff
@@ -176,20 +158,23 @@ int weightToWeightFlags(long edgePointer, double weight) {
                     " nodeB " + nodeAccess.getLat(getNodeB(edgePointer)) + "," + nodeAccess.getLon(getNodeB(edgePointer)));
             weight = MIN_WEIGHT;
         }
-        if (weight > MAX_WEIGHT)
-            weightInt = MAX_WEIGHT_31;
+        if (weight >= WEIGHT_INFINITY)
+            weightInt = (int) WEIGHT_LONG_INFINITY; // negative
         else
-            weightInt = ((int) Math.round(weight * WEIGHT_FACTOR)) << 2;
-        return weightInt;
+            weightInt = (int) Math.round(weight * WEIGHT_FACTOR);
+
+        shortcuts.setInt(edgePointer + S_WEIGHT, weightInt);
     }
 
     double getShortcutWeight(long edgePointer) {
-        // no need for reverseFlags call (shortcut has identical weight if both dies) and also no need for 64bit
-        long flags32bit = getShortcutFlags(edgePointer);
-        double weight = (flags32bit >>> 2) / WEIGHT_FACTOR;
-        if (weight >= MAX_WEIGHT)
+        // might be negative after converting into long -> we need to remove this sign via "& 0xFFFFFFFF". The L is necessary or prepend 8 zeros.
+        long weightLong = (long) shortcuts.getInt(edgePointer + S_WEIGHT) & 0xFFFFFFFFL;
+        if (weightLong == WEIGHT_LONG_INFINITY)
             return Double.POSITIVE_INFINITY;
 
+        double weight = weightLong / WEIGHT_FACTOR;
+        if (weight >= WEIGHT_INFINITY)
+            throw new IllegalArgumentException("too large shortcut weight " + weight + " should get infinity marker bits " + WEIGHT_LONG_INFINITY);
```

Comment:
Should we also add edge information?

Reply:
Not sure. It wouldn't help much. And last but not least: it should never happen ;)